zipkin-storage/elasticsearch/src/main/java/zipkin2/elasticsearch/ElasticsearchStorage.java

```diff
@@ -383,171 +295,25 @@ String buildUrl(IndexTemplates templates, String type) {
-        // the request to an endpoint anyways. This will generally only be when the server is
-        // starting.
-        .orElse(new StaticEndpointGroup(originalEndpoints));
-      EndpointGroupRegistry.register(
-        "elasticsearch", withFallback, EndpointSelectionStrategy.ROUND_ROBIN);
-      clientUrl = urls.get(0).getProtocol() + "://group:elasticsearch" + urls.get(0).getPath();
-    } else {
-      // Just one non-domain URL, can connect directly without enabling load balancing.
-      clientUrl = hosts.get(0);
-    }
-
-    ClientOptionsBuilder options = new ClientOptionsBuilder()
-      .decorator(HttpDecodingClient.newDecorator());
-
-    if (httpLogging() != null) {
-      LoggingClientBuilder loggingBuilder = new LoggingClientBuilder()
-        .requestLogLevel(LogLevel.INFO)
-        .successfulResponseLogLevel(LogLevel.INFO);
-      switch (httpLogging()) {
-        case HEADERS:
-          loggingBuilder.contentSanitizer(unused -> "");
-          break;
-        case BASIC:
-          loggingBuilder.contentSanitizer(unused -> "");
-          loggingBuilder.headersSanitizer(unused -> HttpHeaders.of());
-          break;
-        case BODY:
-        default:
-          break;
-      }
-      options.decorator(loggingBuilder.newDecorator());
-      if (httpLogging() == HttpLoggingLevel.BODY) {
-        options.decorator(RawContentLoggingClient.newDecorator());
-      }
-    }
-
-    clientCustomizer().accept(options);
-    HttpClientBuilder client = new HttpClientBuilder(clientUrl)
-      .factory(clientFactory())
-      .options(options.build());
-
-    return client.build();
-  }
-
   @Override public final String toString() {
-    return "ElasticsearchStorage{hosts=" + hostsSupplier().get()
+    return "ElasticsearchStorage{httpClient=" + lazyHttpClient()
       + ", index=" + indexNameFormatter().index() + "}";
   }
 
-  static boolean isIpAddress(String address) {
-    return zipkin2.Endpoint.newBuilder().parseIp(address);
-  }
-
-  // TODO(anuraaga): Move this upstream - https://github.com/line/armeria/issues/1897
-  static class CompositeEndpointGroup
-    extends AbstractListenable<List<Endpoint>> implements EndpointGroup {
-
-    final List<EndpointGroup> endpointGroups;
-
-    CompositeEndpointGroup(List<EndpointGroup> endpointGroups) {
-      this.endpointGroups = endpointGroups;
-      for (EndpointGroup group : endpointGroups) {
-        group.addListener(unused -> notifyListeners(endpoints()));
+  static {
+    Internal.instance = new Internal() {
+      @Override public HttpCall.Factory http(ElasticsearchStorage storage) {
+        return storage.http();
       }
-    }
+    };
+  }
 
-    @Override public List<Endpoint> endpoints() {
-      List<Endpoint> merged = new ArrayList<>();
-      for (EndpointGroup group : endpointGroups) {
-        merged.addAll(group.endpoints());
-      }
-      return merged;
-    }
+  @Memoized HttpCall.Factory http() {
```

Comment:
fyi I noticed we leaked this public before (though we never cut a release yet). that's what the Internal wrapping is about